core/string/src/lib.rs

```diff
@@ -1144,3 +1145,348 @@ impl ToStringEscaped for [u16] {
     }
 }
+/// Inner elements represented for `RawJsString`.
+pub trait JsStringData: std::fmt::Display + Copy {}
+
+impl JsStringData for u8 {}
+impl JsStringData for u16 {}
+
+/// A mutable builder to create instance of `JsString`.
+///
+/// # Examples
+///
+/// ```rust
+/// use boa_string::JsStringBuilder;
+/// let mut s = JsStringBuilder::new();
+/// s.push(b'x');
+/// s.extend_from_slice(&[b'1', b'2', b'3']);
+/// s.extend([b'1', b'2', b'3']);
+/// let js_string = s.build();
+/// ```
+#[derive(Debug)]
+pub struct JsStringBuilder<T: JsStringData> {
+    cap: usize,
+    len: usize,
+    inner: NonNull<RawJsString>,
+    phantom_data: PhantomData<T>,
+}
+
+impl<T: JsStringData> Clone for JsStringBuilder<T> {
+    #[must_use]
+    fn clone(&self) -> Self {
+        let mut builder = Self::with_capacity(self.capacity());
+        // SAFETY:
+        // - `inner` must be a valid pointer, since it comes from a `NonNull`
+        // allocated above with the capacity of `s`, and initialize to `s.len()` in
+        // ptr::copy_to_non_overlapping below.
+        unsafe {
+            builder
+                .inner
+                .as_ptr()
+                .cast::<u8>()
+                .copy_from_nonoverlapping(self.inner.as_ptr().cast(), self.initialized_bits());
+
+            builder.set_len(self.len());
+        }
+        builder
+    }
+}
+
+impl<T: JsStringData> Default for JsStringBuilder<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: JsStringData> JsStringBuilder<T> {
+    const DATA_SIZE: usize = std::mem::size_of::<T>();
+    const MIN_NON_ZERO_CAP: usize = 8 / Self::DATA_SIZE;
+    const IS_LATIN: bool = Self::DATA_SIZE == std::mem::size_of::<u8>();
+
+    /// Constant default for [`JsStringBuilder`]
+    pub const NEW: Self = Self {
+        cap: 0,
+        len: 0,
+        inner: NonNull::dangling(),
+        phantom_data: PhantomData,
+    };
+
+    /// Create a new `JsStringBuilder` with capacity of zero.
+    #[inline]
+    #[must_use]
+    pub const fn new() -> Self {
+        Self::NEW
+    }
+
+    /// Returns the number of elements in the inner `RawJsString`, also referred to
+    #[inline]
+    #[must_use]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Forces the length of the [`JsStringBuilder`] to `new_len`.
+    ///
+    /// # Safety
+    ///
+    /// - `new_len` must be less than or equal to `capacity()`.
+    /// - The elements at `old_len..new_len` must be initialized.
+    ///
+    #[inline]
+    pub unsafe fn set_len(&mut self, new_len: usize) {
+        debug_assert!(new_len <= self.capacity());
+
+        self.len = new_len;
+    }
+
+    /// Returns the total number of elements the inner `RawJsString` can hold without reallocating
+    #[inline]
+    #[must_use]
+    pub fn capacity(&self) -> usize {
+        self.cap
+    }
+
+    /// Returns the initalized bits of inner `RawJsString`.
+    #[must_use]
+    fn initialized_bits(&self) -> usize {
```

Comment:
I think renaming this to something like `allocated_bytes_len()` would be a better name, feel free to suggest another name if you think it's not accurate.

Reply:
Done.